README.md
[![Build Status](https://travis-ci.org/QuantumBA/pyverless.svg?branch=master)](https://travis-ci.org/QuantumBA/pyverless)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/pyverless.svg)](https://pypi.python.org/pypi/pyverless/)
[![PyPI license](https://img.shields.io/pypi/l/pyverless.svg)](https://pypi.python.org/pypi/pyverless/)
[![PyPI status](https://img.shields.io/pypi/status/pyverless.svg)](https://pypi.python.org/pypi/pyverless/)
# Pyverless

Developing complex APIs within AWS lambdas can be somewhat of a messy task. Lambdas are independent functions that have to work together in order to create a full-blown app, like atoms to a complex organism.

In order to define the infrastructure you may use a framework like [Serverless](https://serverless.com/), but you may find yourself copying and pasting blobs of code within your handler functions, namely for authentication, data validation, error handling and response creation to name a few.

**Enter Pyverless**

Pyverless is a mini-framework with a bunch of utilities that aims to help you create APIs using AWS Lambdas fast and in a consistent way. Pyverless provides the following.

- Class-Based Handlers
- Serializers
- Authentication handling
- JWT and cryptography
- Exceptions
- Configuration
- Warmup handling

Bring more consistency and development speed to your lambda-based APIs!

## Class-Based Handlers

Class based handlers (CBH) use the approach of Django's Class-Based Views to provide code reuse, consistency and generally abstract simple and common tasks. The aim of class-based handlers is to suit a wide range of applications by providing generic Handler classes and mixins.

Within AWS Lambda, a handler is a function that takes an event and a context and returns a response.

Generic CBH are based off the following base handler

### BaseHandler

This class provides the `as_handler()` method that returns a handler function (taking `event` and `context` as arguments).

Usage:

```python
class MyHandler(BaseHandler):
    pass

_myHandler = MyHandler.as_handler()
```

There is a set of generic CBHs to handle basic CRUD operations within an API:

### CreateHandler
Handler that reads the request body and creates the object with each (key, value) pair as a parameter for the constructor.

The `model` attribute must be set on the handler and the `create_object` method can be overwritten.

Usage:

```python
class UserCreateHandler(CreateHandler):

    model = MyUserClass # MyUserClass(k1=v1, k2=v2, ...) for each k,v on body
    required_body_keys = ['email', 'password']
```

### RetrieveHandler
Handler that returns a serialized Object.

The `model` attribute must be set and `id` must be present on the pathParameters.

The user must overwrite either the `serializer` attribute or the `serialize` method.

Usage:

```python
class UserRetrieveHandler(RetrieveHandler):

    model = MyUserClass
    serializer = serialize_user
```

### UpdateHandler
Handler that sets self.object and for each (key, value) pair of the body
sets self.object.key = value.

The `model` attribute must be set and `id` must be present on the pathParameters.

Returns the serialized node and sets the HTTP status code to 200

Usage:

```python
class UserUpdateHandler(UpdateHandler):
    model = MyUserClass
    required_body_keys = ['title', 'body']
    serializer = serialize_user
```


### DeleteHandler
Handler that sets self.object, calls its delete() method and sets the HTTP status code to 204.

The `model` attribute must be set and `id` must be present on the pathParameters.

The user can also overwrite the `get_queryset` method to limit the search.

Usage:

```python
class UserDeleteHandler(DeleteHandler):
    model = MyUserClass
```
### ListHandler
Handler that returns a list of serialized nodes and sets the HTTP status code to 200.

The `model` attribute must be set and the user must overwrite either the `serializer` attribute
or the `serialize` method.

```python
class UserListHandler(ListHandler):
    model = MyUserClass
    serializer = user_serializer
    
    def get_queryset(self):
        return only_some_users
```

## Mixins
There are also a set of **mixins** available:

### RequestBodyMixin

This mixin provides the `get_body()` method which is in charge of gathering the request body dictionary. Define `required_body_keys` and `optinal_body_keys` as follows. Within the handler, you can access the body via `self.body` or by calling `get_body()`

```python
class MyHandler(RequestBodyMixin, BaseHandler):
    required_body_keys = ['name', 'email']
    optinal_body_keys = ['phone']

_myHandler = MyHandler.as_handler()
```

### AuthorizationMixin

This mixin provides the `get_user()` method in charge of getting the user out of an authenticated API call.
Within the handler, you can access the body via `self.user` or by calling `get_user()`. The user will be a object
of the class specified on pyverless settings as `USER_MODEL`.

### RequestBodyMixin

This mixin provides the `get_object()` method in charge of gathering a particular object,
you can access the object via `self.object`.
The `id` of the object will be taken from the pathParameters and
the user must set the `model` attribute on the handler.

### ListMixin

This mixin provides the `get_queryset()` method in charge of getting a list of objects,
you can access the list via `self.queryset`. The user must set the `model` attribute
and either the `serializer` attribute or `serialize()` method on the handler.

### S3FileMixin

This mixin provides the `get_file()` and `get_message_part()` methods in charge of
reading an event from aws S3, you can access the file via `self.file`.

The file will be a `dict()` with the following keys: bucket, owner, file_name, size.

***Warning: Only tested with objectCreated!!!!***

### SQSMessagesMixin

This mixin provides the `get_messages()` method in charge of reading an SQS event from aws.
You can access the list of messages via `self.messages`.

Each message will be a `dict()` with the following keys: attributes, text_message, queue_source, region.

## Serializers

**TODO**


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
### Added

### Changed

### Fixed

### Deprecated

### Removed

## [0.0.51] - 2023-03-14
### Added

### Changed

### Fixed
- Add warmup decorator to handler

### Deprecated

### Removed


## [0.0.49] - 2023-01-10
### Added
- Added logging function attribute to ApiGatewayHandler so multiple logger functions can be used. 
### Changed

### Fixed

### Deprecated

### Removed

## [0.0.48] - 2022-10-27
### Added

### Changed

### Fixed
- Add logging of uncontrolled errors in api gateway

### Deprecated

### Removed


## [0.0.47] - 2022-10-20
### Added

### Changed

### Fixed
- Fix internal server error response

### Deprecated

### Removed

## [0.0.41] - 2022-01-21
### Added
- Add changelog
- Add input serialization

### Changed

### Fixed

### Deprecated

### Removed


pyverless/handlers.py
import json
import logging
import traceback
from typing import Union, Any
import base64

import sentry_sdk
from sentry_sdk import capture_exception, configure_scope

from pyverless.config import settings
from pyverless.decorators import warmup
from pyverless.models import get_user_model
from pyverless.exceptions import BadRequest, Unauthorized, NotFound


class RequestBodyMixin:
    """
    Implement the get_body method that will be called to set self.body as the body
    of the request.
    """

    required_body_keys = []
    optional_body_keys = []

    def get_required_body_keys(self):
        return self.required_body_keys

    def get_body(self):

        body = {}
        missing_keys = []

        try:
            request_body = (
                json.loads(self.event["body"]) if self.event.get("body") else {}
            )
            # The next line is necessary because json.loads('null') = None.
            # 'null' may be a possible value of 'body'
            request_body = request_body if request_body else {}
            if not isinstance(request_body, dict) and not self.get_required_body_keys():
                return request_body
        except json.decoder.JSONDecodeError:
            message = "Malformed body"
            print(f"MALFORMED: {self.event.get('body')}")
            self.error = (message, 400)
            raise BadRequest(message=message)

        # Collect all required keys and values. Collected missing keys are
        # reported as an error.
        for key in self.get_required_body_keys():
            try:
                body[key] = request_body[key]
            except KeyError:
                missing_keys.append(key)

        if missing_keys:
            message = "Missing key(s): %s" % ", ".join(missing_keys)
            self.error = (message, 400)
            raise BadRequest(message=message)

        # Collect all optional keys and values.
        for key in self.optional_body_keys:
            try:
                body[key] = request_body[key]
            except KeyError:
                pass

        return body


class QueryParamsMixin:
    """
    Implement the get_queryparams method that will be called to populate self.queryparams
    """

    required_query_keys = []

    required_multivalue = []
    optional_multivalue = []

    def get_queryparams(self):
        missing_keys = set()

        queryparams = self.event.get("queryStringParameters") or {}
        multivalue_queryparams = self.event.get("multiValueQueryStringParameters") or {}

        result = queryparams

        for key in self.required_query_keys:
            if key in queryparams:
                result[key] = queryparams[key]
            else:
                missing_keys.add(key)
        for key in self.required_multivalue:
            if key in multivalue_queryparams:
                result[key] = multivalue_queryparams[key]
            else:
                missing_keys.add(key)

        if missing_keys:
            message = "Missing key(s): %s" % ", ".join(missing_keys)
            self.error = (message, 400)
            raise BadRequest(message=message)

        # Collect all optional keys and values.
        for key in self.optional_multivalue:
            if key in multivalue_queryparams:
                result[key] = multivalue_queryparams[key]

        return result


class SQSMessagesMixin:
    """
    Implement the get_messages method that will be called to set self.messages

    Each message is a dict with the following keys:

    - attributes
    - text_message
    - queue_source
    - region
    """

    def get_messages(self):
        messages = []

        try:
            temp_messages = self.event["Records"] if self.event["Records"] else []
        except json.decoder.JSONDecodeError:
            message = "Malformed message"
            self.error = (message, 400)
            raise BadRequest(message=message)

        for message in temp_messages:
            temp_message = {}

            temp_message["attributes"] = self.get_message_part(
                message, "messageAttributes", message["messageId"]
            )
            temp_message["text_message"] = self.get_message_part(
                message, "body", message["messageId"]
            )
            temp_message["queue_source"] = message["eventSourceARN"]
            temp_message["region"] = message["awsRegion"]

            messages.append(temp_message)

        return messages

    def get_message_part(self, message, part, message_id):

        try:
            message_part = message[part] if message[part] else {}
        except json.decoder.JSONDecodeError:
            response_text = "Malformed {} in message {}".format(part, message_id)
            self.error = (response_text, 400)
            raise BadRequest(message=response_text)

        return message_part


class S3FileMixin:
    """
    Implement the get_file method that will be called to set self.file

    The file is a dict with the following keys:

    - event_name
    - bucket
    - owner
    - file_name
    - size
    """

    def get_file(self):
        try:
            temp_file_event = self.event["Records"] if self.event["Records"] else []
        except json.decoder.JSONDecodeError:
            message = "Malformed message"
            self.error = (message, 400)
            raise BadRequest(message=message)

        file = {
            "event_name": temp_file_event[0]["eventName"],
            "bucket": temp_file_event[0]["s3"]["bucket"]["name"],
            "owner": temp_file_event[0]["s3"]["bucket"]["ownerIdentity"]["principalId"],
            "file_name": temp_file_event[0]["s3"]["object"]["key"],
        }
        # on ObjectRemoved event, the size is not present
        if temp_file_event[0]["s3"]["object"].get("size"):
            file["size"] = temp_file_event[0]["s3"]["object"]["size"]

        return file

    def get_message_part(self, message, part, message_id):

        try:
            message_part = message[part] if message[part] else {}
        except json.decoder.JSONDecodeError:
            response_text = "Malformed {} in message {}".format(part, message_id)
            self.error = (response_text, 400)
            raise BadRequest(message=response_text)

        return message_part


class AuthorizationMixin:
    """
    Implement the get_user method that will be called to set self.user

    USER_MODEL must be set on the pyverless settings
    """

    def get_user(self):
        # TODO: if there is no token or it is invalid it should return 401 instead of 403
        # see: https://stackoverflow.com/questions/50143518/401-unauthorized-vs-403-forbidden-which-is-the-right-status-code-for-when-the-u?rq=1
        try:
            user_id = self.event["requestContext"]["authorizer"]["principalId"]
        except KeyError:
            self.error = ("Unauthorized", 403)
            raise Unauthorized()

        UserModel = get_user_model()
        user = getattr(UserModel, settings.MODEL_MANAGER).get_or_none(uid=user_id)

        if user:
            return user
        else:
            self.error = ("Unauthorized", 403)
            raise Unauthorized()


class ObjectMixin:
    """
    Implement the get_object method that will be called to set self.object,
    the object 'id' must be present on the pathParameters.

    The user can also overwrite the get_queryset method to limit the visibility.

    The 'model' attribute must be set on the handler.
    """

    model = None
    serializer = None
    id_in_path = "id"

    def get_object(self):
        object_id = self.event["pathParameters"][self.id_in_path]

        # When get_queryset is overriden by the user, the queryset it returns may
        # be a list. Is such case, the list has to be filtered to get the
        # desired object.
        if isinstance(self.queryset, list):

            def filt(instance):
                return instance.uid == object_id

            filtered = list(filter(filt, self.queryset))

            obj = filtered.pop() if filtered else None
        else:
            obj = self.queryset.get_or_none(uid=object_id)

        if not obj:
            self.error = ("Resource Not Found", 404)
            raise NotFound()

        return obj

    def get_queryset(self):
        return getattr(self.model, settings.MODEL_MANAGER)

    def serialize(self, instance):
        return self.serializer(instance=instance).data


class ListMixin:
    """
    Implement the get_queryset method that will be called to set self.queryset

    The 'model' attribute must be set on the handler and the user must overwrite
    either the 'serializer' attribute or the 'serialize' method.
    """

    model = None
    serializer = None

    def get_queryset(self):
        return getattr(self.model, settings.MODEL_MANAGER)

    def serialize(self, instance):
        return self.serializer(instance=instance).data


class BaseHandler:

    # type hints
    user: Any
    queryset: Any
    object: Any
    body: Any
    queryparams: dict
    messages: list
    file: dict

    event: dict
    error: Union[list, tuple]
    force_error: bool
    headers: dict
    is_base64: bool

    success_code = 200

    def perform_action(self):
        """
        This method is to be overriden. Here is where the particular handler
        action is performed. Its return value is a dictionary with the response body.
        """
        return self.response_body

    @classmethod
    def as_handler(cls):
        """
        Returns a lambda handler function.
        """

        @warmup
        def handler(event, context):
            self = cls()

            self.event = event
            self.context = context
            self.error = None
            self.user = None
            self.body = {}
            self.response_body = {}
            self.messages = {}
            self.headers = {}
            self.force_error = False
            self.is_base64 = False

            # set user, queryset, object, body and response_body (that is, if the handler
            # uses the apropiate mixin and the method is avaliable)
            pairs = [
                ("body", "get_body"),
                ("queryparams", "get_queryparams"),
                ("user", "get_user"),
                ("queryset", "get_queryset"),
                ("object", "get_object"),
                ("messages", "get_messages"),
                ("file", "get_file"),
                ("response_body", "perform_action"),
            ]

            for attr, method in pairs:
                if hasattr(self, method):
                    try:
                        setattr(self, attr, getattr(self, method)())
                    except Exception as e:
                        if not self.error:
                            tb = traceback.format_exc()
                            return self.render_500_error_response(e, tb)
                if self.error:
                    return self.render_error_response(
                        self.error[0],
                        self.error[1],
                        self.error[2] if len(self.error) == 3 else None,
                    )

            return self.render_response(self.response_body, self.success_code)

        return handler

    def render_response(self, body, status_code):
        """
        Given a body and status_code, returns a dictionary in the format of a valid
        AWS handler response.
        As AWS API Gateway works as a proxy, lambda is in charge of setting and
        returning the applicable CORS headers. Check the next link for more info:
        http://docs.aws.amazon.com/apigateway/latest/developerguide/how-to-cors.html
        """
        response = {
            "statusCode": status_code,
            "body": json.dumps(body)
            if not self.is_base64
            else base64.b64encode(body).decode("utf-8"),
            "headers": {
                "Access-Control-Allow-Origin": settings.CORS_ORIGIN,
                "Access-Control-Allow-Headers": settings.CORS_HEADERS,
                "Access-Control-Allow-Methods": "*",
                **self.headers,
            },
            "isBase64Encoded": self.is_base64,
        }
        return response

    def render_error_response(self, message, status_code, field=None):
        """
        Given a message and error status_code, returns a dictionary in the format of a valid
        AWS handler response with our error body definition. This allows us to be consistent.
        """
        body = {"code": status_code, "message": message}
        if field:
            body["field"] = field
        if not self.force_error:
            response = self.render_response(body, status_code)
        else:
            raise Exception(json.dumps(body))

        return response

    def render_500_error_response(self, e, tb):
        """
        Given a exception and traceback, log unhandled exceptions using the python
        logger and/or sentry. Also, create an http response with status code 500
        and message.
        """

        # Log errors using the python logger.
        logger = logging.getLogger()
        logger.setLevel(logging.ERROR)

        logger.exception(e)

        # Log errors in sentry if USE_SENTRY setting variable is set to True.
        if settings.USE_SENTRY:
            sentry_sdk.init(dsn=settings.SENTRY_DNS)

            with configure_scope() as scope:

                if self.user:
                    scope.user = {"id": self.user.uid, "email": self.user.email}

                scope.set_tag("stage", settings.STAGE)
                scope.set_extra("class", self.__class__)
                scope.set_extra("body", self.body)
                scope.set_extra("event", self.event)

            capture_exception(e)

        # Send an 500 error response with traceback and event information in the
        # response body if DEBUG setting variable is set to True.
        if settings.DEBUG:
            message = {"traceback": tb, "event": self.event}
        else:
            message = "Internal Server Error"

        return self.render_error_response(message, 500)


class ReadSQSHandler(SQSMessagesMixin, BaseHandler):
    """
    Handler that returns a list of SQS messages and sets the HTTP status code to 200.

    Each message is a dict with the following keys:

    - attributes
    - text_message
    - queue_source
    - region
    """

    success_code = 200

    def perform_action(self):
        return self.messages


class CreateHandler(RequestBodyMixin, BaseHandler):
    """
    Handler that reads the request body and creates the object with each (key, value) pair
    as the params for the constructor.

    The 'model' attribute must be set on the handler and the user can overwrite the 'create_object' method.
    """

    success_code = 201

    def create_object(self):
        obj = self.model(**self.body).save()

        return obj

    def perform_action(self):
        obj = self.create_object()

        return obj.uid


class RetrieveHandler(ObjectMixin, BaseHandler):
    """
    Handler that returns a serialized Object.

    The 'model' attribute must be set and 'id' must be present on the pathParameters.

    The user also has to define the 'serialize' method on the handler.
    """

    success_code = 200

    def perform_action(self):
        data = self.serialize(self.object) if self.object else None

        return data


class ListHandler(ListMixin, QueryParamsMixin, BaseHandler):
    """
    Handler that returns a list of serialized nodes and sets the HTTP status code to 200.

    The 'model' attribute must be set and the user must overwrite
    either the 'serializer' attribute or the 'serialize' method.
    """

    success_code = 200
    limit = None
    optional_query_keys = ["offset", "limit"]

    def perform_action(self):
        _list = []
        offset = int(self.queryparams.get("offset", 0))
        limit = self.queryparams.get("limit") or self.limit
        end = None
        if limit is not None:
            end = offset + int(limit)

        for obj in self.queryset[offset:end]:
            _list.append(self.serialize(obj))
        return _list


class UpdateHandler(RequestBodyMixin, ObjectMixin, BaseHandler):
    """
    Handler that sets self.object and for each (key, value) pair of the body
    sets self.object.key = value.

    The 'model' attribute must be set and 'id' must be present on the pathParameters.

    Returns the serialized node and sets the HTTP status code to 200
    """

    success_code = 200

    def perform_action(self):

        for key, value in self.body.items():
            setattr(self.object, key, value)

        self.object.save()

        return self.serialize(self.object)


class DeleteHandler(ObjectMixin, BaseHandler):
    """
    Handler that sets self.object, calls its delete() method and sets the HTTP status code to 204.

    The 'model' attribute must be set and 'id' must be present on the pathParameters.

    The user can also overwrite the 'get_queryset' method to limit the search.
    """

    success_code = 204

    def perform_action(self):
        self.object.delete()

        return {}
